packages/compose/src/higher-order/with-safe-timeout/index.tsx

```diff
@@ -0,0 +1,84 @@
+/**
+ * External dependencies
+ */
+import { without } from 'lodash';
+// eslint-disable-next-line no-restricted-imports
+import type { ComponentType } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import createHigherOrderComponent from '../../utils/create-higher-order-component';
+// eslint-disable-next-line no-duplicate-imports
+import type { PropInjectingHigherOrderComponent } from '../../utils/create-higher-order-component';
+
+/**
+ * We cannot use the `Window['setTimeout']` and `Window['clearTimeout']`
+ * types here because those functions include functionality that is not handled
+ * by this component, like the ability to pass extra arguments.
+ *
+ * In the case of this component, we only handle the simplest case where
+ * `setTimeout` only accepts a function (not a string) and an optional delay.
+ */
+type TimeoutProps = {
+	setTimeout: ( fn: () => void, delay: number ) => number;
+	clearTimeout: ( id: number ) => void;
+};
+
+/**
+ * A higher-order component used to provide and manage delayed function calls
+ * that ought to be bound to a component's lifecycle.
+ */
+const withSafeTimeout: PropInjectingHigherOrderComponent< TimeoutProps > = createHigherOrderComponent(
+	< TProps extends TimeoutProps >(
+		OriginalComponent: ComponentType< TProps >
+	) => {
+		return class WrappedComponent extends Component<
+			Omit< TProps, keyof TimeoutProps >
+		> {
+			timeouts: number[];
+
+			constructor( props ) {
```

Comment:
TypeScript doesn't seem to like this line: `Parameter 'props' implicitly has an 'any' type`

Is this enough?

```suggestion
			constructor( props: any ) {
```